### ISSUE TITLE
xfrd: bad transfer 0 from x - nsd 4.3.1

### ISSUE THREAD

**Issue by jaredmauch** [NLnetLabs/nsd#104]
I recently upgraded to 4.3.1 and since then some zones are transferring very often, in 1 second intervals while logging errors about bad transfer 0 

```
# nsd-control zonestatus reklama.lv
zone:	reklama.lv
	state: refreshing
	served-serial: "2020053001 since 2020-06-02T12:48:29"
	commit-serial: "2020053001 since 2020-06-02T12:48:29"
	wait: "1 sec between attempts"
```

With the appropriate logs as well:

```
Jun  7 02:49:09 puck nsd[3508867]: [2020-06-07 02:49:09.019] nsd[3508867]: info: xfrd: zone reklama.lv ignoring old serial from A.B.C.D
Jun  7 02:49:09 puck nsd[3508867]: [2020-06-07 02:49:09.019] nsd[3508867]: info: xfrd: zone reklama.lv bad transfer 0 from A.B.C.D
Jun  7 02:49:09 puck nsd[3508867]: xfrd: zone reklama.lv bad transfer 0 from A.B.C.D
Jun  7 02:49:10 puck nsd[3508867]: xfrd: zone reklama.lv ignoring old serial from A.B.C.D
Jun  7 02:49:10 puck nsd[3508867]: [2020-06-07 02:49:10.275] nsd[3508867]: info: xfrd: zone reklama.lv ignoring old serial from A.B.C.D
Jun  7 02:49:10 puck nsd[3508867]: [2020-06-07 02:49:10.275] nsd[3508867]: info: xfrd: zone reklama.lv bad transfer 0 from A.B.C.D
Jun  7 02:49:10 puck nsd[3508867]: xfrd: zone reklama.lv bad transfer 0 from A.B.C.D
Jun  7 02:49:11 puck nsd[3508867]: xfrd: zone reklama.lv ignoring old serial from A.B.C.D
```

I also have done a pcap of a transfer and have that available, and at least one of the servers that is logging this (there are many) is returning this:

```
version.bind.		0	CH	TXT	"9.8.2rc1-RedHat-9.8.2-0.68.rc1.el6_10.7"
```

Does someone want some PCAPs to look closer, or have an idea of what might be causing this?

**Comment by jaredmauch** [comment]
Looks like doing a `force_transfer` resolve this, but the 1 second wait between attempts does look to be an issue.

```
Jun  7 13:30:19 puck nsd[3508867]: [2020-06-07 13:30:19.676] nsd[3508867]: info: xfrd: zone discns.ca ignoring old serial from A.B.C.D
Jun  7 13:30:19 puck nsd[3508867]: xfrd: zone discns.ca bad transfer 0 from A.B.C.D
Jun  7 13:30:19 puck nsd[3508867]: [2020-06-07 13:30:19.676] nsd[3508867]: info: xfrd: zone discns.ca bad transfer 0 from A.B.C.D
Jun  7 13:30:20 puck nsd[3508867]: xfrd: zone discns.ca ignoring old serial from A.B.C.D
Jun  7 13:30:20 puck nsd[3508867]: [2020-06-07 13:30:20.719] nsd[3508867]: info: xfrd: zone discns.ca ignoring old serial from A.B.C.D
Jun  7 13:30:20 puck nsd[3508867]: xfrd: zone discns.ca bad transfer 0 from A.B.C.D
Jun  7 13:30:20 puck nsd[3508867]: [2020-06-07 13:30:20.719] nsd[3508867]: info: xfrd: zone discns.ca bad transfer 0 from A.B.C.D
Jun  7 13:30:21 puck nsd[3508867]: [2020-06-07 13:30:21.656] nsd[3508867]: info: control cmd:  force_transfer discns.ca
Jun  7 13:30:21 puck nsd[3508867]: control cmd:  force_transfer discns.ca
Jun  7 13:30:21 puck nsd[3508867]: xfrd: zone discns.ca written received XFR packet from A.B.C.D with serial 2020060601 to disk
Jun  7 13:30:21 puck nsd[3508867]: [2020-06-07 13:30:21.694] nsd[3508867]: info: xfrd: zone discns.ca written received XFR packet from A.B.C.D with serial 2020060601 to disk
Jun  7 13:30:21 puck nsd[3508867]: [2020-06-07 13:30:21.694] nsd[3508867]: info: xfrd: zone discns.ca committed "received update to serial 2020060601 at 2020-06-07T13:30:21 from A.B.C.D"
Jun  7 13:30:21 puck nsd[3508867]: xfrd: zone discns.ca committed "received update to serial 2020060601 at 2020-06-07T13:30:21 from A.B.C.D"
Jun  7 13:30:26 puck nsd[3508880]: zone discns.ca. received update to serial 2020060601 at 2020-06-07T13:30:21 from A.B.C.D of 5614 bytes in 0.000113 seconds
Jun  7 13:30:26 puck nsd[3508880]: [2020-06-07 13:30:26.700] nsd[3508880]: info: zone discns.ca. received update to serial 2020060601 at 2020-06-07T13:30:21 from A.B.C.D of 5614 bytes in 0.000113 seconds
Jun  7 13:30:26 puck nsd[3508867]: zone discns.ca serial 3020644816 is updated to 2020060601
Jun  7 13:30:26 puck nsd[3508867]: [2020-06-07 13:30:26.710] nsd[3508867]: info: zone discns.ca serial 3020644816 is updated to 2020060601
Jun  7 13:30:51 puck nsd[3508880]: writing zone discns.ca to file```

**Comment by jaredmauch** [comment]
It seems once i forced all the zones to update the logs stopped.

It seems the issue is that when this occurs and is rejected due to serial number, the retry timer is 1 second.

I tried setting this on the per-zone, but it doesn't seem to have the expected impact of reducing the transfer attempt frequency from 1s -> 30s.

```
        min-retry-time: 30
        min-refresh-time: 30
```

**Comment by wtoorop** [comment]
@jaredmauch I have been unsuccessfully trying to reproduce, but maybe it is related with: #103 
Can you still reproduce? If so, it would be interesting to see log output from `nsd -L 2 -F 0x0060` with a  nsd configured and compiled with `--enable-checking` option.

**Comment by jaredmauch** [comment]
maybe.. i added in some serial printing/checking so next time it happens I can debug further as well.

**Comment by jaredmauch** [comment]
so I'm seeing what seems to be serial number related issues between the 4.2.4 and 4.3.1 files that seem to be triggering this.

The non-ntohl() values appear to be written to disk for the serial number which results in the SOA serial looking wrong and the transfer failing.

Then there is the secondary issue of the 1 second retry when this occurs.

**Comment by jaredmauch** [comment]
```
; zone wods.ca written by NSD 4.2.4 on Fri May 15 14:06:19 2020
; received update to serial 3020645816 at 2020-05-15T14:06:01 from a.b.c.d
$ORIGIN ca.
wods    86400   IN      SOA     ns1.zuluru.net. greg.zuluru.net. (
                3020645816 86400 7200 3600000 86400 )
```

from my logs:

```
Jun 14 06:18:26 puck nsd[2236557]: xfrd: zone wods.ca ignoring old serial (3020645816/2020060600) from 52.5.16.56
Jun 14 06:18:26 puck nsd[2236557]: xfrd: zone wods.ca bad transfer 0 from a.b.c.d
```

I added in printing of the serials to understand why

**Comment by wtoorop** [comment]
Thanks @jaredmauch . We're on the small (can be 0 even) retry and refresh minimums. A fix will be released shortly, but the no-ntohl() serial number is peculiar! Do you have that for more than one zone?

**Comment by jaredmauch** [comment]
It seems that way, i'm going to add some additional printing and will send patch/PR.

**Comment by jaredmauch** [comment]
yeah checking my logs..

You can see the above is what was written to disk, and i added printing of the serials based on disk.. and here's all my logs of zone serial changes:

```
/var/log/messages:Jun 14 22:19:58 puck nsd[3372816]: axfr for wods.ca. from 192.35.168.144 refused, no acl matches
/var/log/messages:Jun 15 11:05:16 puck nsd[3532192]: axfr for wods.ca. from 51.38.98.191 refused, no acl matches
/var/log/messages:Jun 15 11:10:39 puck nsd[2236557]: control cmd:  force_transfer wods.ca
/var/log/messages:Jun 15 11:10:39 puck nsd[2236557]: xfrd: zone wods.ca written received XFR packet from A.B.C.D with serial 2020060600 to disk
/var/log/messages:Jun 15 11:10:39 puck nsd[2236557]: xfrd: zone wods.ca committed "received update to serial 2020060600 at 2020-06-15T11:10:39 from A.B.C.D"
/var/log/messages:Jun 15 11:10:39 puck nsd[2236582]: zone wods.ca. received update to serial 2020060600 at 2020-06-15T11:10:39 from A.B.C.D of 3362 bytes in 0.000103 seconds
/var/log/messages:Jun 15 11:10:39 puck nsd[2236557]: zone wods.ca serial 3020645816 is updated to 2020060600
/var/log/messages:Jun 15 11:10:49 puck nsd[2236582]: writing zone wods.ca to file /etc/nsd/secondaries/wods.ca
/var/log/messages-20200524:May 18 12:49:18 puck nsd[869]: zone wods.ca read with success
/var/log/messages-20200524:May 18 12:49:18 puck nsd[869]: [2020-05-18 12:49:18.583] nsd[869]: info: zone wods.ca read with success
/var/log/messages-20200531:May 24 00:57:02 puck nsd[1423507]: notify for wods.ca. from A.B.C.D serial 2020052400
/var/log/messages-20200531:May 24 00:57:02 puck nsd[1423507]: [2020-05-24 00:57:02.739] nsd[1423507]: info: notify for wods.ca. from A.B.C.D serial 2020052400
/var/log/messages-20200531:May 24 00:57:02 puck nsd[1423507]: notify for wods.ca. from A.B.C.D serial 2020052400
/var/log/messages-20200531:May 24 00:57:02 puck nsd[1423507]: [2020-05-24 00:57:02.739] nsd[1423507]: info: notify for wods.ca. from A.B.C.D serial 2020052400
/var/log/messages-20200531:May 24 00:57:09 puck nsd[1423530]: notify for wods.ca. from A.B.C.D serial 2020052404
/var/log/messages-20200531:May 24 00:57:09 puck nsd[1423530]: [2020-05-24 00:57:09.407] nsd[1423530]: info: notify for wods.ca. from A.B.C.D serial 2020052404
/var/log/messages-20200531:May 24 00:57:09 puck nsd[1423530]: [2020-05-24 00:57:09.407] nsd[1423530]: info: notify for wods.ca. from A.B.C.D serial 2020052404
/var/log/messages-20200531:May 24 00:57:09 puck nsd[1423530]: notify for wods.ca. from A.B.C.D serial 2020052404
/var/log/messages-20200531:May 24 05:20:26 puck nsd[1460805]: [2020-05-24 05:20:26.473] nsd[1460805]: info: notify for wods.ca. from A.B.C.D serial 2020052405
/var/log/messages-20200531:May 24 05:20:26 puck nsd[1460805]: notify for wods.ca. from A.B.C.D serial 2020052405
/var/log/messages-20200531:May 24 05:20:26 puck nsd[1460805]: notify for wods.ca. from A.B.C.D serial 2020052405
/var/log/messages-20200531:May 24 05:20:26 puck nsd[1460805]: [2020-05-24 05:20:26.973] nsd[1460805]: info: notify for wods.ca. from A.B.C.D serial 2020052405
/var/log/messages-20200531:May 24 05:21:20 puck nsd[1460914]: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200531:May 24 05:21:20 puck nsd[1460914]: [2020-05-24 05:21:20.552] nsd[1460914]: info: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200531:May 24 05:21:20 puck nsd[1460914]: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200531:May 24 05:21:20 puck nsd[1460914]: [2020-05-24 05:21:20.552] nsd[1460914]: info: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200607:Jun  1 06:47:44 puck nsd[3508880]: zone wods.ca read with success
/var/log/messages-20200607:Jun  1 06:47:44 puck nsd[3508880]: [2020-06-01 06:47:44.867] nsd[3508880]: info: zone wods.ca read with success
/var/log/messages-20200607:Jun  5 01:04:39 puck nsd[124068]: [2020-06-05 01:04:39.016] nsd[124068]: info: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200607:Jun  5 01:04:39 puck nsd[124068]: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200607:Jun  5 01:04:41 puck nsd[124087]: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200607:Jun  5 01:04:41 puck nsd[124087]: [2020-06-05 01:04:41.517] nsd[124087]: info: notify for wods.ca. from A.B.C.D serial 2020052406
/var/log/messages-20200607:Jun  6 16:34:15 puck nsd[694584]: [2020-06-06 16:34:15.463] nsd[694584]: info: notify for wods.ca. from A.B.C.D serial 2020060600
/var/log/messages-20200607:Jun  6 16:34:15 puck nsd[694584]: notify for wods.ca. from A.B.C.D serial 2020060600
/var/log/messages-20200607:Jun  6 16:34:18 puck nsd[694583]: [2020-06-06 16:34:18.966] nsd[694583]: info: notify for wods.ca. from A.B.C.D serial 2020060600
/var/log/messages-20200607:Jun  6 16:34:18 puck nsd[694583]: notify for wods.ca. from A.B.C.D serial 2020060600
/var/log/messages-20200614:Jun  8 06:23:39 puck nsd[1169636]: zone wods.ca read with success
/var/log/messages-20200614:Jun  8 06:23:39 puck nsd[1169636]: [2020-06-08 06:23:39.689] nsd[1169636]: info: zone wods.ca read with success
/var/log/messages-20200614:Jun  9 06:26:12 puck nsd[1417389]: zone wods.ca read with success
/var/log/messages-20200614:Jun  9 06:26:12 puck nsd[1417389]: [2020-06-09 06:26:12.053] nsd[1417389]: info: zone wods.ca read with success
/var/log/messages-20200614:Jun 10 07:22:26 puck nsd[1684502]: zone wods.ca read with success
/var/log/messages-20200614:Jun 10 09:42:31 puck nsd[1711689]: zone wods.ca read with success
/var/log/messages-20200614:Jun 10 09:47:35 puck nsd[1712928]: zone wods.ca read with success
/var/log/messages-20200614:Jun 11 18:41:14 puck nsd[2236582]: zone wods.ca read with success
/var/log/messages-20200614:Jun 11 19:14:59 puck nsd[2236557]: control cmd:  force-xfer wods.ca
/var/log/messages-20200614:Jun 11 19:15:03 puck nsd[2236557]: control cmd:  transfer wods.ca
```

**Comment by and0x000** [comment]
I can see pretty much the same behavior with zones having `wait: "1 sec between attempts"` on my NSD 4.3.1. 

The zone was unable to be transferred, as the remote server refused communication. Shouldn't there be some kind of hysteresis, for when the the transfer failed?

I took a peek into the xfrd.c, where it looks like a hysteresis is intended.
https://github.com/NLnetLabs/nsd/blob/master/xfrd.c#L793

From what I observed so far, this hysteresis seems not to kick in.
```
zone:   example.com
        state: refreshing
        served-serial: "2019101501 since 2020-06-15T13:28:04"
        commit-serial: "2019101501 since 2020-06-15T13:28:04"
        wait: "1 sec between attempts"
```

However, if I trigger a `force_transfer` for the zone (while still being unable to communicate with the master), it seems to start using this hysteresis.
```
~ # nsd-control zonestatus example.com
zone:   example.com
        state: refreshing
        served-serial: none
        commit-serial: none
        wait: "34 sec between attempts"
~ # nsd-control zonestatus example.com
zone:   example.com
        state: refreshing
        served-serial: none
        commit-serial: none
        wait: "102 sec between attempts"
```

**Comment by jaredmauch** [comment]
Resolved in #103 and 4.3.2 release
